### PR TITLE
Only build generic Linux packages.

### LINKED ISSUE #36: Rename and improve generic Linux packages.
But if we are making this move I would like to call the package 'linux' and not 'lnx'... but we can leave this for py3.

_Originally posted by @adiroiban in https://github.com/chevah/server/issues/5718#issuecomment-1325377150_

### PR DESCRIPTION
**Pull request by dumol**
Scope
-----

Fixes #36 

Changes
-------

Only build generic glibc and musl Linux packages.   Also, renamed `lnx` and `lnx_musl` packages to `linux` and `linux_musl`.

The generic Linux package is now built on Ubuntu 18.04, to support RHEL 7/8. (For the issues building on CentOS 6/7, see the comments below.)  

~~The macOS package is now built on macOS 11, as 10.15 runners are to be deprecated by GitHub soonish.~~ (Not for now, too many issues.)

Latest `cryptography` is now packaged, using wheels everywhere.

**Drive-by changes**:
  * Updated OpenSSL 1.1.1 version (it's only used for Python's `ssl` on non-Windows platforms now).
  * Updated `pip` and all the Python modules to latest versions (`safety` complained about a couple of them).
  * Removed `subprocess32` module.
  * Forward-ported the clean-ups from https://github.com/chevah/python-package/pull/173 (no `wheel` in the distributable, `include/` is moved to `lib/`, no `ssl/`).

Testing
-------

Please review all changes.

Check the automated tests.

Test the resulting distributables?

### CODE REVIEW

**Comment by adiroiban** [comment]
It's ok if generic Linux are rhel 7 only.
We should merge this as it is

We can worry about rhel 6 or 5 later if any customer needs it

**Comment by dumol** [comment]
As said above, don't get too enthusiastic about CentOS 7, the `/lib64/libfreebl3.so` lib is linked against dynamically when building on it. As seen at https://github.com/chevah/pythia/actions/runs/4415314095/jobs/7739368782.

I've also checked how deep this dep goes, and it's not quite bad… Everything and its mother is linked to it, including the `python` binary. And the `/lib64/libfreebl3.so` lib on CentOS 7 belongs to a package that is not removable, being a dependency of both yum and systemd. Some say it's because of a FIPS requirement.

Basically, this means that a `pythia` package built on CentOS 7 would only work on RHEL 7+ and their derivatives.

**Comment by dumol** [comment]
On a second thought, don't get too enthusiastic about CentOS 6 either… Even without being able to integrate `cryptography` on CentOS 6, I've manually checked deps for the built binaries and libs. Unfortunately, the `/lib64/libfreebl3.so` is linked against there too, and the `nss-softokn-freebl` package is already a system one on CentOS 6.

Any other suggestions for a minimalistic distro with long-term supported versions? To support RHEL 7, it would have to be based on glibc 2.19 or older (e.g. Ubuntu 12.04). To support RHEL 6, it needs glibc 2.12 or older (e.g. Ubuntu 10.04).

**Comment by adiroiban** [comment]
OK. Then let's go with just RHEL 8 and 9. We should be fine for a start.

Thanks

**Review by adiroiban — APPROVED** [review]
Thanks for the changes. I am a bit worried about the usage of OpenSSL v3 via the latest cryptography 

but maybe that is fine.

I you want,
I think that you can go ahead and merge this and later we can look at how to handle the openssl version differences.

----------

As the first step in SFTPPlus migration to py3 I would still prefer to use OpenSSL v1.1.1.

but is not a big deal.

I am aproving it so that we can continue with the py3.10 PR

File: .github/workflows/bare.yaml [hunk 1]
```diff
@@ -55,7 +35,7 @@ jobs:
       fail-fast: false
       matrix:
         runs-on: [ windows-2022, windows-2019 ]
```
[the remark this answers is not in the document]
Comment (dumol) [reply .github/workflows/bare.yaml:37]: No, we don't need that. But I would like to keep `x86` support as long as there's no extra work for us.

It's extra work for GHA, but Pythia is an open-source project, so we get free GHA here. More so, others might need an `x86` Windows Python 3.x runtime packaged in this way…

File: .github/workflows/bare.yaml [hunk 2]
```diff
@@ -13,26 +13,6 @@ on:
   pull_request:
     branches: [ main ]
 
-  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (dumol) [reply .github/workflows/bare.yaml:16]: Sorry, it doesn't make much sense here, where you can't run the tests without running the build.

File: .github/workflows/bare.yaml
[the remark this answers is not in the document]
Comment (dumol) [reply .github/workflows/bare.yaml:1]: Done at https://github.com/chevah/pythia/pull/38/commits/d5c69c05a028c6c95fff5d10af0b93d701fcf06b.

File: README.rst [hunk 3]
```diff
@@ -22,21 +22,16 @@ Supported platforms
...
-* Ubuntu Server 18.04 and 20.04
-* Alpine Linux
-
 Platforms that should work, but are not regularly tested:
 
+* all musl-based Linux distributions (musl 1.1.24+)
 * FreeBSD 12 and newer
 * OpenBSD 6.7 and newer
-* Solaris 11.4.
+* Solaris 11.4 (x86 only).
```
[the remark this answers is not in the document]
Comment (dumol) [reply README.rst:34]: I've actually removed support for AIX, HP-UX, and Solaris SPARC when starting Pythia. Solaris x86 was working at the time with minimum effort.

Haven't checked on it since, but it should be little work, if needed. As this is an open-source project, I would rather keep this as is, where Solaris x86 is listed under _Platforms that should work, but are not regularly tested_.

File: build.conf [hunk 4]
```diff
@@ -13,28 +13,28 @@ XZ_VERSION="5.2.5"
...
+SETUPTOOLS_VERSION="67.6.0"
 
 # Python modules versions to be used everywhere possible.
-CFFI_VERSION="1.15.0"
-CRYPTOGRAPHY_VERSION="3.4.8"
-SUBPROCESS32_VERSION="3.5.4"
-BCRYPT_VERSION="3.2.2"
-PSUTIL_VERSION="5.9.0"
-SETPROCTITLE_VERSION="1.2.3"
+# Latest cryptography/bcrypt require Rust, use only wheels for them.
```
[the remark this answers is not in the document]
Comment (dumol) [reply build.conf:26]: The comment documents that latest `cryptography` and `bcrypt` modules requires a Rust compiler for building. Therefore, we need to use wheels for these modules everywhere: Windows, Linux, macOS.

It's true that we still build Python 1.1.1x on all non-Windows platforms. It's only used for Python's `ssl` (as documented in the PR text). The same is true for the Pythia packages on Windows, where we don't build Python ourselves. The upstream Python runtimes we repackage here also use OpenSSL 1.1.1x.

I don't have any strong preferences regarding this, but I would rather follow the upstream Python packagers. Therefore, switch to OpenSSL 3.0.x for Python's `ssl` when upstream Python is also packaged with OpenSSL 3.0.x. Either way, preferably in a dedicated PR.

**Comment by dumol** [comment]
> Thanks for the changes. I am a bit worried about the usage of OpenSSL v3 via the latest cryptography

Why worry about this?

> 
> but maybe that is fine.
> 
> I you want, I think that you can go ahead and merge this and later we can look at how to handle the openssl version differences.
> 
> As the first step in SFTPPlus migration to py3 I would still prefer to use OpenSSL v1.1.1.

Not possible unless we build `cryptography` ourselves on **all** platforms again. (Which is not worth the trouble since Rust is required.)

 
> but is not a big deal.
> 
> I am aproving it so that we can continue with the py3.10 PR

Thanks! Close the discussions satisfactorily answered, and I'll proceed with the merge into master when there are no outstanding problems to discuss…

**Comment by dumol** [comment]
As discussed in private on Slack, merging this…
